File: Aulas/Aula5/Agregacao_exemplo.java
import java.ArrayList;

// Classe Endereco
class Endereco {
    private String rua;
    private String cidade;
    private String estado;
    private String cep;

    public Endereco(String rua, String cidade, String estado, String cep) {
        this.rua = rua;
        this.cidade = cidade;
        this.estado = estado;
        this.cep = cep;
    }

    public void mostrarEndereco() {
        System.out.println("Rua: " + rua + ", Cidade: " + cidade + ", Estado: " + estado + ", CEP: " + cep);
    }
}

// Classe Pessoa
class Pessoa {
    private String nome;
    private int idade;
    private Endereco endereco;

    public Pessoa(String nome, int idade) {
        this.nome = nome;
        this.idade = idade;
        this.endereco = null;  // Inicialmente sem endereço
    }

    public void adicionarEndereco(Endereco endereco) {
        this.endereco = endereco;
    }

    public void mostrarInformacoes() {
        System.out.println("Nome: " + nome + ", Idade: " + idade);
        if (endereco != null) {
            System.out.println("Endereço:");
            endereco.mostrarEndereco();
        } else {
            System.out.println("Endereço não disponível");
        }
    }

    public String getNome() {
        return nome;
    }
}

// Classe Empresa
class Empresa {
    private String nome;
    private String cnpj;
    private List<Pessoa> funcionarios;

    public Empresa(String nome, String cnpj) {
        this.nome = nome;
        this.cnpj = cnpj;
        this.funcionarios = new ArrayList<>();
    }

    public void contratarFuncionario(Pessoa funcionario) {
        funcionarios.add(funcionario);
    }

    public void listarFuncionarios() {
        System.out.println("Funcionários da empresa " + nome + ":");
        for (Pessoa funcionario : funcionarios) {
            System.out.println(funcionario.getNome());
        }
    }
}

// Cliente
public class Agregacao_exemplo {
    public static void main(String[] args) {
        Endereco endereco1 = new Endereco("Av. Principal", "Cidade A", "Estado X", "12345-678");
        Pessoa pessoa1 = new Pessoa("João", 30);
        pessoa1.adicionarEndereco(endereco1);

        Endereco endereco2 = new Endereco("Rua Secundária", "Cidade B", "Estado Y", "98765-432");
        Pessoa pessoa2 = new Pessoa("Maria", 25);
        pessoa2.adicionarEndereco(endereco2);

        Empresa empresa = new Empresa("Empresa ABC", "12345678901234");
        empresa.contratarFuncionario(pessoa1);
        empresa.contratarFuncionario(pessoa2);

        empresa.listarFuncionarios();
    }
}

File: Aulas/Aula5/Agregacao_exemplo.go
package main

import "fmt"

// Struct Endereco
type Endereco struct {
    rua    string
    cidade string
    estado string
    cep    string
}

func (e Endereco) MostrarEndereco() {
    fmt.Printf("Rua: %s, Cidade: %s, Estado: %s, CEP: %s\n", e.rua, e.cidade, e.estado, e.cep)
}

// Struct Pessoa
type Pessoa struct {
    nome     string
    idade    int
    endereco *Endereco
}

func NewPessoa(nome string, idade int) *Pessoa {
    return &Pessoa{nome: nome, idade: idade, endereco: nil}
}

func (p *Pessoa) AdicionarEndereco(endereco Endereco) {
    p.endereco = &endereco
}

func (p Pessoa) MostrarInformacoes() {
    fmt.Printf("Nome: %s, Idade: %d\n", p.nome, p.idade)
    if p.endereco != nil {
        fmt.Println("Endereço:")
        p.endereco.MostrarEndereco()
    } else {
        fmt.Println("Endereço não disponível")
    }
}

// Struct Empresa
type Empresa struct {
    nome        string
    cnpj        string
    funcionarios []*Pessoa
}

func NewEmpresa(nome string, cnpj string) *Empresa {
    return &Empresa{nome: nome, cnpj: cnpj, funcionarios: []*Pessoa{}}
}

func (e *Empresa) ContratarFuncionario(funcionario *Pessoa) {
    e.funcionarios = append(e.funcionarios, funcionario)
}

func (e Empresa) ListarFuncionarios() {
    fmt.Printf("Funcionários da empresa %s:\n", e.nome)
    for _, funcionario := range e.funcionarios {
        fmt.Println(funcionario.nome)
    }
}

// Cliente
func main() {
    endereco1 := Endereco{"Av. Principal", "Cidade A", "Estado X", "12345-678"}
    pessoa1 := NewPessoa("João", 30)
    pessoa1.AdicionarEndereco(endereco1)

    endereco2 := Endereco{"Rua Secundária", "Cidade B", "Estado Y", "98765-432"}
    pessoa2 := NewPessoa("Maria", 25)
    pessoa2.AdicionarEndereco(endereco2)

    empresa := NewEmpresa("Empresa ABC", "12345678901234")
    empresa.ContratarFuncionario(pessoa1)
    empresa.ContratarFuncionario(pessoa2)

    empresa.ListarFuncionarios()
}
